File: src/statistics.R
library(pROC)

source('src/constants.R')

print_interactions_and_main_effects <- function(cv_fit, preprocessed_data) {
    # Extracting the index where lambdaHat1Std equals lambda
    i_1Std <- which(cv_fit$lambdaHat1Std == cv_fit$lambda)
    coefs <- coef(cv_fit$glinternetFit)[[i_1Std]]

    # Processing main effects
    main_effects <- coefs$mainEffects$cont
    cat('Main effects:\n')
    for (effect in main_effects) {
        col_index <- effect
        print(colnames(preprocessed_data$X_train[, ..col_index]))
    }

    # Processing interactions
    cat('\nInteractions:\n')
    if (length(coefs$interactions$catcont) > 0) {
        for (row in 1:nrow(coefs$interactions$catcont)) {
            var_idx <- (coefs$interactions$catcont[row, 2])
            print(colnames(preprocessed_data$X_train[, ..var_idx]))
        }
    }
}

compute_and_print_roc_auc <- function(cv_fit, preprocessed_data) {

    X_test <- preprocessed_data$X_test
    y_test <- preprocessed_data$y_test

    predictions <- as.vector(predict(cv_fit, X_test, type = "response"))

    roc_curve <- suppressMessages(roc(y_test, predictions, quietly = TRUE))
    auc_score <- auc(roc_curve)
    cat('\nAUC score: ', auc_score)
    cat('\n\n####################################################################################################')
    cat('\n####################################################################################################\n')

    # Return the AUC score
    return(auc_score)
}

cases_summary <- function(pheno) {

    results <- list()

    for (disease_name in names(DISEASE_CODES)) {
        disease_code <- DISEASE_CODES[disease_name]

        s_asian_cases_1       <- sum(pheno$population == "s_asian" & pheno[[disease_code]] == 2, na.rm = TRUE)
        white_british_cases_1 <- sum(pheno$population == "white_british" & pheno[[disease_code]] == 2, na.rm = TRUE)
        s_asian_cases_0       <- sum(pheno$population == "s_asian" & pheno[[disease_code]] == 1, na.rm = TRUE)
        white_british_cases_0 <- sum(pheno$population == "white_british" & pheno[[disease_code]] == 1, na.rm = TRUE)

        # Count SA cases with disease in each final_split category
        s_asian_cases_train <- sum(
            pheno$population == "s_asian" 
            & pheno[[disease_code]] == 2 
            & pheno$final_split == "train", 
            na.rm = TRUE)
        
        s_asian_cases_val   <- sum(
            pheno$population == "s_asian"
            & pheno[[disease_code]] == 2 
            & pheno$final_split == "val", 
            na.rm = TRUE)
        
        s_asian_cases_test  <- sum(
            pheno$population == "s_asian" 
            & pheno[[disease_code]] == 2 
            & pheno$final_split == "test", 
            na.rm = TRUE)

        results[[disease_name]] <- c(
            SA_1 = s_asian_cases_1, WB_1 = white_british_cases_1, 
            SA_0 = s_asian_cases_0, WB_0 = white_british_cases_0,
            SA_Train = s_asian_cases_train, SA_Val = s_asian_cases_val, SA_Test = s_asian_cases_test)
    }

    for (disease_name in names(results)) {
        disease_code <- DISEASE_CODES[disease_name]
        cat("Disease Name:", disease_name, "\nDisease Code:", disease_code, "\n")

        disease_matrix <- matrix(unlist(results[[disease_name]][1:4]), nrow = 2, byrow = TRUE, 
                                 dimnames = list(c("SA", "WB"), c("Disease 1", "Disease 0")))
        print(disease_matrix)

        # Calculate and print the ratio of SA 1 to WB 1 cases
        ratio_sa_wb <- ifelse(disease_matrix[2,1] > 0, disease_matrix[1,1] / disease_matrix[2,1], "NA (No WB cases)")
        cat("Ratio of SA 1 to WB 1 cases:", ratio_sa_wb, "\n")

        # Print the number of SA cases with disease in each final_split category
        cat("SA cases with disease in Train:", results[[disease_name]]["SA_Train"], "\n")
        cat("SA cases with disease in Val:", results[[disease_name]]["SA_Val"], "\n")
        cat("SA cases with disease in Test:", results[[disease_name]]["SA_Test"], "\n\n")
    }
        }

File: src/statistics.py
"""
Disease Cases Summary Module.

This module provides a function to summarize disease cases based on phenotypic data. 
It calculates the number of cases for different populations and disease statuses, 
and prints a summary including the ratio of cases between populations and the distribution of cases across different data splits.
"""

import pandas as pd

from constants import DISEASE_CODES

def cases_summary(pheno):
    """
    Summarizes disease cases based on phenotype DataFrame.

    Parameters:
    - pheno (DataFrame): Phenotype DataFrame containing population, disease codes, and data splits.

    Prints:
    - Summary of disease cases by population and data split.
    - Ratio of cases between South Asian and White British populations.
    """
    results = {}

    for disease_name, disease_code in DISEASE_CODES.items():
        counts = {}
        counts['SA_1'] = pheno[(pheno['population'] == 's_asian') & (pheno[disease_code] == 2)].shape[0]
        counts['WB_1'] = pheno[(pheno['population'] == 'white_british') & (pheno[disease_code] == 2)].shape[0]
        counts['SA_0'] = pheno[(pheno['population'] == 's_asian') & (pheno[disease_code] == 1)].shape[0]
        counts['WB_0'] = pheno[(pheno['population'] == 'white_british') & (pheno[disease_code] == 1)].shape[0]
        counts['SA_Train'] = pheno[(pheno['population'] == 's_asian') & (pheno[disease_code] == 2) & (pheno['final_split'] == 'train')].shape[0]
        counts['SA_Val'] = pheno[(pheno['population'] == 's_asian') & (pheno[disease_code] == 2) & (pheno['final_split'] == 'val')].shape[0]
        counts['SA_Test'] = pheno[(pheno['population'] == 's_asian') & (pheno[disease_code] == 2) & (pheno['final_split'] == 'test')].shape[0]

        results[disease_name] = counts

    for disease_name, counts in results.items():
        print(f"Disease Name: {disease_name}, Disease Code: {DISEASE_CODES[disease_name]}")
        disease_matrix = pd.DataFrame({'Disease 1': [counts['SA_1'], counts['WB_1']], 'Disease 0': [counts['SA_0'], counts['WB_0']]},
                                      index=['SA', 'WB'])
        print(disease_matrix)

        ratio_sa_wb = "NA (No WB cases)" if disease_matrix.loc['WB', 'Disease 1'] == 0 else disease_matrix.loc['SA', 'Disease 1'] / disease_matrix.loc['WB', 'Disease 1']
        print(f"Ratio of SA 1 to WB 1 cases: {ratio_sa_wb}")
        print(f"SA cases with disease in Train: {counts['SA_Train']}")
        print(f"SA cases with disease in Val: {counts['SA_Val']}")
        print(f"SA cases with disease in Test: {counts['SA_Test']}\n")
        